File: iOS/depolllib/depolllib/analytics/Analytics.swift
//
//  Analytics.swift
//  depolllib
//
//  Created by David Poll on 6/16/21.
//  Copyright © 2021 DepollSoft. All rights reserved.
//

import Foundation

public class Analytics {
    public static let appOpenEvent = "app_open"
    private static let endpoint = URL(string: "https://api.depollsoft.xyz/analytics")!
    private var prefix: String
    public init(prefix: String) {
        self.prefix = prefix
    }
    
    private func persistKey(_ key: String) -> String {
        return "\(prefix).\(key)"
    }
    
    private func getOSInfo() -> String {
        let os = ProcessInfo().operatingSystemVersion
        return String(os.majorVersion) + "." + String(os.minorVersion) + "." + String(os.patchVersion)
    }
    
    private lazy var modelName: String = {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
        return identifier
    }()
    
    private func startOfHour(_ date: Date) -> Date {
        let timestamp = date.timeIntervalSince1970
        let startOfHour = timestamp - timestamp.truncatingRemainder(dividingBy: 3600)
        return Date(timeIntervalSince1970: startOfHour)
    }
    
    private func updateTags(_ tags: Set<String>, forEvent event: String) -> Set<String> {
        var newTags = tags
        let dailyKey = persistKey("\(event).daily")
        let hourlyKey = persistKey("\(event).hourly")
        let now = Date()
        let lastDaily = UserDefaults.standard.object(forKey: dailyKey) as? Date
        let lastHourly = UserDefaults.standard.object(forKey: hourlyKey) as? Date
        let calendar = Calendar.current
        if lastDaily == nil || calendar.startOfDay(for: lastDaily!) != calendar.startOfDay(for: now) {
            newTags.insert("daily")
            UserDefaults.standard.set(now, forKey: dailyKey)
        }
        if lastHourly == nil || startOfHour(lastHourly!) != startOfHour(now) {
            newTags.insert("hourly")
            UserDefaults.standard.set(now, forKey: hourlyKey)
        }
        return newTags
    }
    
    private func formatDate(_ date: Date) -> String{
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "YYYY-MM-dd HH:mm:ss.SSSSSS"
        return formatter.string(from: date)
    }
    
    private func transformDictionary<T>(_ dict: [String: T], keyKey: String, valueKey: String) -> [[String: Any]] {
        var result: [[String: Any]] = []
        for (key, value) in dict {
            result.append([
                            keyKey: key,
                            valueKey:value
            ])
        }
        return result
    }
    
    public func logEvent(_ eventName: String,
                         tags: Set<String> = [],
                         fields: [String: String] = [:],
                         metrics: [String: Double] = [:]) {
        do{
            var request = URLRequest(url: Analytics.endpoint)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: [
                "tags": Array(updateTags(tags, forEvent: eventName)),
                "app_id": Bundle.main.bundleIdentifier!,
                "app_version": Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion")!,
                "event_timestamp": formatDate(Date()),
                "event_timezone": TimeZone.current.identifier,
                "event_name": eventName,
                "fields": transformDictionary(fields, keyKey: "key", valueKey: "value"),
                "metrics": transformDictionary(metrics, keyKey: "name", valueKey: "value"),
                "device_model": modelName,
                "platform": "iOS",
                "platform_version": getOSInfo()
            ])
            URLSession.shared.dataTask(with: request).resume()
        } catch {
            print("Caught while logging: \(error)")
        }
    }
    
    public private(set) static var sharedInstance: Analytics = Analytics(prefix: "depolllib.analytics")
}

File: Android/depollsoft.lib.kotlin/src/main/java/depollsoft/lib/analytics/Analytics.kt
package depollsoft.lib.analytics

import android.content.Context
import android.os.Build
import android.util.Log
import androidx.core.content.pm.PackageInfoCompat
import depollsoft.lib.activity.RichApplication
import depollsoft.lib.toJsonElement
import depollsoft.lib.toJsonString
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

class Analytics(appContext: Context, val sharedPrefs: String) {
    private val prefs = appContext.getSharedPreferences(sharedPrefs, Context.MODE_PRIVATE)
    private fun startOfHour(date: Date): Date {
        val timestamp = date.time
        val startOfHour = timestamp - timestamp % MILLIS_PER_HOUR
        return Date(startOfHour)
    }

    private fun startOfDay(date: Date): Date {
        val timestamp = date.time
        val startOfDay = timestamp - timestamp % MILLIS_PER_DAY
        return Date(startOfDay)
    }

    private fun updateTags(tags: Set<String>, event: String): Set<String> {
        synchronized(prefs) {
            var newTags = tags.toMutableSet()
            val dailyKey = "${event}.daily"
            val hourlyKey = "${event}.hourly"
            val now = Date()
            val lastDaily = Date(prefs.getLong(dailyKey, 0L))
            val lastHourly = Date(prefs.getLong(hourlyKey, 0L))
            val editorLazy = lazy { prefs.edit() }
            val editor by editorLazy
            if (startOfDay(lastDaily) != startOfDay(now)) {
                newTags.add("daily")
                editor.putLong(dailyKey, now.time)
            }
            if (startOfHour(lastHourly) != startOfHour(now)) {
                newTags.add("hourly")
                editor.putLong(hourlyKey, now.time)
            }
            if (editorLazy.isInitialized()) {
                editor.apply()
            }
            return newTags
        }
    }

    private fun <T> transformDictionary(
        dict: Map<String, T>,
        keyKey: String,
        valueKey: String
    ): List<Map<String, Any?>> =
        dict.map { mapOf(keyKey to it.key, valueKey to it.value) }

    public fun logEvent(
        eventName: String,
        tags: Set<String> = emptySet(),
        fields: Map<String, String> = emptyMap(),
        metrics: Map<String, Double> = emptyMap()
    ) {
        val event = mapOf(
            "tags" to updateTags(tags, eventName),
            "app_id" to appContext.packageName,
            "app_version" to PackageInfoCompat.getLongVersionCode(
                appContext.packageManager.getPackageInfo(appContext.packageName, 0)
            ),
            "event_timestamp" to dateFormat.format(Date()),
            "event_timezone" to TimeZone.getDefault().id,
            "event_name" to eventName,
            "fields" to transformDictionary(fields, "key", "value"),
            "metrics" to transformDictionary(metrics, "name", "value"),
            "device_model" to Build.MODEL,
            "platform" to "android",
            "platform_version" to Build.VERSION.SDK_INT
        )

        CoroutineScope(Dispatchers.IO + Job()).launch {
            val eventString = event.toJsonString()

            val connection = endpoint.openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                connection.addRequestProperty("Content-Type", "application/json")
                connection.doOutput = true
                connection.outputStream.bufferedWriter().use {
                    it.write(eventString)
                }
                if (connection.responseCode < 200 || connection.responseCode >= 400) {
                    Log.e(
                        "depollsoft.lib",
                        "Sending analytics failed: ${connection.responseCode} ${connection.responseMessage}"
                    )
                }
            } catch(e: Exception) {
                // Definitely don't fail during analytics collection
            } finally {
                connection.disconnect()
            }
        }
    }

    companion object {
        const val APP_OPEN = "app_open"
        private const val MILLIS_PER_HOUR = 1000 * 60 * 60
        private const val MILLIS_PER_DAY = MILLIS_PER_HOUR * 24
        private val endpoint = URL("https://api.depollsoft.xyz/analytics")
        private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSSSSS", Locale.US)

        lateinit var appContext: Context
        private fun ctx(): Context = if (this::appContext.isInitialized) appContext else RichApplication.getAppContext()

        val default by lazy { Analytics(ctx(), "depollsoft.lib.analytics") }
    }
}